### DISCUSSION TITLE
3D geometric numpy array to networkx - similar to geographic examples but for more general approach

### DISCUSSION THREAD

**Discussion by Sh4zKh4n** [networkx/networkx#4540] [Q&A]
***edit to rephrase question

I have noticed the really interesting examples you have been creating in the new docs for dev for GeoSpatial. I wondered though would you be able to do a more general version for ndim numpy arrays? For example, i can't get my head around how to build a graph from a 3d numpy array representing voxels (one thats been labelled) or for one that is a point cloud which is either a data frame or x, y, z, label (so four columns as either a data frame or a numpy array). 

Often when images are segmented the 0 label is just empty space so redundant, and it is redundant space. So I have thought about building a point cloud from this data and someone pointed out the following as an option to convert this to a point cloud (which I think is almost the same as what the sparse library does).

``` python
import gradslam as gs
import numpy as np

# create basic voxel grid
voxel_grid = np.zeros((10, 10, 10, 1), dtype=np.uint8)
voxel_grid[4:6, 4:6, 4:6] = 100

# convert voxel grid to xyz coordinates and features
nonzero_inds = np.nonzero(voxel_grid)[:-1]
points = torch.tensor(nonzero_inds, dtype=torch.float32).T
features = torch.tensor(voxel_grid[nonzero_inds])

# construct gradslam.Pointclouds object
pointclouds = gs.Pointclouds(points=[points], features=[features])

# visualize
o3d.visualization.draw_geometries([pointclouds.open3d(0)])
```

I think with the sparse Library it gives out a column based sparse matrix of the form

dim 0, dim 1 , dim 2, data
z         y          x          feature(label)

So now I have a point cloud, with a position and vector and would say like to create a graph network where nodes are connected to a nearest neighbour (nodes are labels with (point coordinates and with the label) and the edges are either with a weight 1 or a weight determined by the Euclidean distance between the nodes within a given distance (i think the geospatial examples all use external libraries that work in 2d space but not in higher dims such as 3d). It sounds like it isn't much difference between the examples in the new geospatial examples and could be beneficial for creating pixel or voxel based representations. ( I should say my example here is simple but the data I want to represent is as a voxel array, too big for in memory and I am using dask to chunk and analyse, so this is a simplistic example here)

The numpy array I have is actually a chunked dask.array of shape (13,000, 5000, 10,000). Alot of it is empty space in an 8 bit file, for thats alot of nodes but if change this to a ndim sparse array from an array using the sparse array it appears thats actually going to be almost euivalent to the above

So I get that one could iterate through the array and get the value for (i,j,k) by conver

```python
import numpy as np
import dask.array as da
import networkx as nx
import dask.dataframe as dd
import pandas as pd

big_array =da.random.random((15000, 2500, 10000), chunks=(1000, 500, 1000), dtype=np.uint8) # big array 
big_array[big_array<0.9] = 0 # fill most of the array with zeros

ba_s = sparse.COO(big_array)


bas_df = dd.io.from_dask_array(ba_s, columns=['z', 'y', 'x', 'label'])

bas_df.to_hdf('df_sparse_BA.hdf', '/data')    

bas_df = dd.read_hdf('df_sparse_BA.hdf') 

def graph(df, object_id=object_id):
    # create kdtree
    tree = spatial.cKDTree(df[['z', 'y', 'x']])

    # get neighbours within distance for every point, store in dataframe as edges
    edges = pd.DataFrame({'src':[], 'tgt':[]}, dtype=int)
    for source, target in enumerate(tree.query_ball_tree(tree, r=2)):
        target.remove(source)
        if target:
            edges = edges.append(pd.DataFrame({'src':[source] * len(target), 'tgt':target}), ignore_index=True)

    # create graph for points using edges from Balltree query
    G = nx.from_pandas_dataframe(edges, 'src', 'tgt')

    for i in sorted(G.nodes()):
        G.node[i]['label'] = nodes.label[i]
        G.node[i]['x'] = nodes.x[i]
        G.node[i]['y'] = nodes.y[i]
        G.node[i]['z'] = nodes.z[i]
   return G
```

i am not sure this would work for a big graph with networkx but I wonder if this is the right way?  

Would be really interested to know how you think this could be solved, since it seems very similar to the geospatial methods.
